opensearch/src/http/middleware.rs

```diff
@@ -0,0 +1,128 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+use std::{clone::Clone, fmt::Debug, future, io::Write, ops::Deref};
+
+use base64::{prelude::BASE64_STANDARD, write::EncoderWriter as Base64Encoder};
+use reqwest::{ClientBuilder, Identity, Request, RequestBuilder};
+
+#[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+use crate::auth::ClientCertificate;
+use crate::{
+    auth::Credentials,
+    future::BoxFuture,
+    http::headers::{HeaderValue, AUTHORIZATION},
+    BoxError,
+};
+
+use super::transport::BuildError;
+
+pub trait Middleware: Send + Sync {
+    fn configure_client(&self, client_builder: ClientBuilder) -> Result<ClientBuilder, BuildError> {
+        Ok(client_builder)
+    }
+
+    fn pre_process_request(&self, request_builder: RequestBuilder) -> RequestBuilder {
+        request_builder
+    }
+
+    fn post_process_request(
+        &self,
+        request: Request,
+    ) -> BoxFuture<Result<Request, BoxError<'static>>> {
+        Box::pin(future::ready(Ok(request)))
+    }
```

Comment:
The naming seems off. Is there some kind of Rust standard for middleware/callbacks? I would adopt either an event style `on_before`, `on_after` or a stack style `configure` and `call`?